toptal/picasso
# [FX-NULL] Restore number input behaviour and look

### Why is it important

Without this fix, portals can not be updated to latest Picasso due to visual and form submission flow changes.

After this fix, the Client Portal has all checks passing https://github.com/toptal/client-portal/pull/9995 ✅ 

### Description

When `NumberInput` is used from [corresponding form control](https://picasso.toptal.net/?path=/story/forms-numberinput--numberinput), the form control passes `min` and `max` (after https://github.com/toptal/picasso/pull/4610) to `NumberInput`.

`NumberInput` passes `min` and `max` further to actual number HTML input which leads to 
- the width of control is determined by browser based on `min` and `max` values (https://stackoverflow.com/questions/33283901/input-number-max-attribute-resizes-field)
- form validates the input field on submit (https://developer.mozilla.org/en-US/docs/Web/API/ValidityState/stepMismatch)

**Behaviour introduced in this pull request**

This pull request gets back to the original behaviour when https://github.com/toptal/picasso/tree/master/packages/base/NumberInput had empty (or invalid, which has the same effect) `min` and `max` attributes that was preventing browser from modifying the look and behaviour of the component.

### How to test


- [Temploy](https://picasso.toptal.net/FX-NULL-do-not-pass-min-and-max)
- The https://github.com/toptal/client-portal/pull/9995 should
    - have no visual changes linked to number input
    - act and look the same way as staging, please check [the pull request description](https://github.com/toptal/client-portal/pull/9995) for details
- Use the snippet below to test
    - form submission
    - how arrows work regarding `min` and `max` values with different `step`

```typescript
import React from 'react'
import { FormActionsContainer } from '@toptal/picasso'
import { SPACING_4 } from '@toptal/picasso-utils'
import {
  FormNonCompound as Form,
  NumberInput,
  SubmitButton,
} from '@toptal/picasso-forms'

const Example = () => {
  return (
    <Form
      autoComplete='off'
      onSubmit={values => window.alert(JSON.stringify(values, undefined, 2))}
      initialValues={{}}
    >
      <NumberInput
        required
        min={1}
        step={5}
        name='default-age'
        label="What's your age?"
        placeholder='e.g. 25'
      />
      <FormActionsContainer top={SPACING_4}>
        <SubmitButton>Submit</SubmitButton>
      </FormActionsContainer>
    </Form>
  )
}

export default Example
```

### Screenshots

Before


https://github.com/user-attachments/assets/d8b07d14-ca2e-4633-9ec8-8b90befcb28a


After

https://github.com/user-attachments/assets/49c55c3d-4e45-4b79-901b-c595f3852463


### Development checks

- [N/A] Double check if `picasso-tailwind-merge` requires major update (check its `README.md`)
- [N/A] Make sure that additions and changes on the design follow [Toptal's BASE design](https://design.toptal.net/), and it's been already discussed with designers at #-base-core
- [N/A] Annotate all `props` in component with documentation
- [N/A] Create `examples` for component


> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>
